File: language/runtime/ceylon/language/Iterable.java
package ceylon.language;

import com.redhat.ceylon.compiler.java.metadata.Annotation;
import com.redhat.ceylon.compiler.java.metadata.Annotations;
import com.redhat.ceylon.compiler.java.metadata.Ceylon;
import com.redhat.ceylon.compiler.java.metadata.Name;
import com.redhat.ceylon.compiler.java.metadata.SatisfiedTypes;
import com.redhat.ceylon.compiler.java.metadata.TypeInfo;
import com.redhat.ceylon.compiler.java.metadata.TypeParameter;
import com.redhat.ceylon.compiler.java.metadata.TypeParameters;
import com.redhat.ceylon.compiler.java.metadata.Variance;

@Ceylon(major = 4)
@TypeParameters({@TypeParameter(value = "Element", variance = Variance.OUT),
                 @TypeParameter(value = "Absent", variance = Variance.OUT, 
                     defaultValue = "ceylon.language::Null", 
                     satisfies = "ceylon.language::Null")})
@SatisfiedTypes("ceylon.language::Container<Element,Absent>")
public interface Iterable<Element,Absent> extends Container<Element,Absent> {

    @Annotations(@Annotation("default"))
    public long getSize();

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @Override
    public boolean getEmpty();
    
    @Annotations(@Annotation("formal"))
    @TypeInfo("ceylon.language::Iterator<Element>")
    public Iterator<? extends Element> getIterator();

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @TypeInfo("Absent|Element")
    public Element getFirst();

    @Annotations({@Annotation("actual"), @Annotation("default")})
    @TypeInfo("Absent|Element")
    public Element getLast();

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Iterable<Element, Null>")
    public Iterable<? extends Element, ? extends java.lang.Object> getRest();

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Sequential<Element>")
    public Sequential<? extends Element> getSequence();

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Iterable<Result,ceylon.language::Null>")
    @TypeParameters(@TypeParameter("Result"))
    public <Result> Iterable<? extends Result, ? extends java.lang.Object> map(@Name("collecting")
        @TypeInfo("ceylon.language::Callable<Result,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Result> collecting);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Iterable<Element,ceylon.language::Null>")
    public Iterable<? extends Element, ? extends java.lang.Object> filter(@Name("selecting")
        @TypeInfo("ceylon.language::Callable<ceylon.language::Boolean,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Boolean> selecting);

    @Annotations(@Annotation("default"))
    @TypeInfo("Result")
    @TypeParameters(@TypeParameter("Result"))
    public <Result> Result fold(@Name("initial")
		@TypeInfo("Result") Result initial,
		@Name("accumulating")
        @TypeInfo("ceylon.language::Callable<Result,ceylon.language::Tuple<Result|Element,Result,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>>")
        Callable<? extends Result> accumulating);

    @Annotations(@Annotation("default"))
    @TypeInfo("Element|ceylon.language::Null")
    public Element find(@Name("selecting")
        @TypeInfo("ceylon.language::Callable<ceylon.language::Boolean,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Boolean> selecting);

    @Annotations(@Annotation("default"))
    @TypeInfo("Element|ceylon.language::Null")
    public Element findLast(@Name("selecting")
        @TypeInfo("ceylon.language::Callable<ceylon.language::Boolean,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Boolean> selecting);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Sequential<Element>")
    public Sequential<? extends Element> sort(@Name("comparing")
        @TypeInfo("ceylon.language::Callable<ceylon.language::Null|ceylon.language::Comparison,ceylon.language::Tuple<Element,Element,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>>")
        Callable<? extends Comparison> comparing);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Sequential<Result>")
    @TypeParameters(@TypeParameter("Result"))
    public <Result> Sequential<? extends Result> collect(@Name("collecting")
        @TypeInfo("ceylon.language::Callable<Result,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Result> collecting);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Sequential<Element>")
    public Sequential<? extends Element> select(@Name("selecting")
        @TypeInfo("ceylon.language::Callable<ceylon.language::Boolean,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Boolean> selecting);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Boolean")
    public boolean any(@Name("selecting")
        @TypeInfo("ceylon.language::Callable<ceylon.language::Boolean,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Boolean> selecting);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Boolean")
    public boolean every(@Name("selecting")
        @TypeInfo("ceylon.language::Callable<ceylon.language::Boolean,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Boolean> selecting);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Iterable<Element,ceylon.language::Null>")
    public Iterable<? extends Element, ? extends java.lang.Object> skipping(@Name("skip")
        @TypeInfo("ceylon.language::Integer")
        long skip);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Iterable<Element,ceylon.language::Null>")
    public Iterable<? extends Element, ? extends java.lang.Object> taking(@Name("take")
        @TypeInfo("ceylon.language::Integer")
        long take);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Iterable<Element,ceylon.language::Null>")
    public Iterable<? extends Element, ? extends java.lang.Object> by(@Name("step")
        @TypeInfo("ceylon.language::Integer")
        long step);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Integer")
    public long count(@Name("selecting")
        @TypeInfo("ceylon.language::Callable<ceylon.language::Boolean,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
        Callable<? extends Boolean> selecting);

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Iterable<Element&ceylon.language::Object,Absent>")
    public Iterable<? extends Element,? extends Absent> getCoalesced();

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Iterable<ceylon.language::Entry<ceylon.language::Integer,Element&ceylon.language::Object>,Absent>")
    public Iterable<? extends Entry<? extends Integer, ? extends Element>, ? extends Absent> getIndexed();

    @Annotations(@Annotation("default"))
    @TypeInfo("ceylon.language::Iterable<Element|Other,ceylon.language::Null>")
    @TypeParameters(@TypeParameter("Other"))
    public <Other> Iterable chain(@Name("other")
            @TypeInfo("ceylon.language::Iterable<Other,ceylon.language::Null>")
            Iterable<? extends Other, ? extends java.lang.Object> other);

    @Annotations(@Annotation("default"))
    @TypeParameters(@TypeParameter(value="Grouping",satisfies="ceylon.language::Object"))
    @TypeInfo("ceylon.language::Map<Grouping,ceylon.language::Sequence<Element>>")
    public <Key> Map<? extends Key, ? extends Sequence<? extends Element>> group(@Name("grouping")
            @TypeInfo("ceylon.language::Callable<Grouping,ceylon.language::Tuple<Element,Element,ceylon.language::Empty>>")
            Callable<? extends Key> grouping);

    @Override
    @Annotations({@Annotation("actual"), @Annotation("default")})
    public boolean contains(@Name("element") java.lang.Object element);

}

File: compiler-js/src/main/js/iterable.js
function initTypeProtoI(a, b, c){} //IGNORE
function initTypeProto(a,b,c,d){}//IGNORE
function initType(a,b){}//IGNORE
function Basic(x){}//IGNORE
function ArraySequence(x){}//IGNORE
function Comprehension(x){}//IGNORE
function Exception(x){}//IGNORE
function String$(x){}//IGNORE
var Container,$finished,empty,larger,smaller,exports;//IGNORE

function Iterable(wat) {
    return wat;
}
initTypeProtoI(Iterable, 'ceylon.language::Iterable', $init$ContainerWithFirstElement());
function $init$Iterable() { return Iterable; }
var Iterable$proto=Iterable.$$.prototype;
Iterable$proto.getEmpty = function() {
    return this.getIterator().next() === $finished;
}
Iterable$proto.getFirst = function() {
    var e = this.getIterator().next();
    return e === $finished ? null : e;
}
Iterable$proto.getRest = function() {
    return this.skipping(1);
}
Iterable$proto.getSequence = function() {
    var a = [];
    var iter = this.getIterator();
    var next;
    while ((next = iter.next()) !== $finished) {
        a.push(next);
    }
    return ArraySequence(a, this.$$targs$$);
}
Iterable$proto.$map = function(mapper) {
    var iter = this;
    return Comprehension(function() {
        var it = iter.getIterator();
        return function() {
            var e = it.next();
            if(e !== $finished) {return mapper(e);}
            return $finished;
        }
    });
}
Iterable$proto.$filter = function(select) {
    var iter = this;
    return Comprehension(function() {
        var it = iter.getIterator();
        return function() {
            do {
                var e = it.next();
            } while ((e !== $finished) && !select(e));
            return e;
        }
    });
}
Iterable$proto.fold = function(ini, accum) {
    var r = ini;
    var iter = this.getIterator();
    var e; while ((e = iter.next()) !== $finished) {
        r = accum(r, e);
    }
    return r;
}
Iterable$proto.find = function(select) {
    var iter = this.getIterator();
    var e; while ((e = iter.next()) !== $finished) {
        if (select(e)) {
            return e;
        }
    }
    return null;
}
Iterable$proto.findLast = function(select) {
    var iter = this.getIterator();
    var last = null;
    var e; while ((e = iter.next()) !== $finished) {
        if (select(e)) {
            last = e;
        }
    }
    return last;
}
Iterable$proto.$sort = function(/*Callable<Comparison?,Element,Element>*/comparing) {
    var a = [];
    var iter = this.getIterator();
    var e; while ((e = iter.next()) !== $finished) {
        a.push(e);
    }
    a.sort(function(x,y) {
        var r = comparing(x,y);
        if (r === larger) return 1;
        if (r === smaller) return -1;
        return 0;
    });
    return ArraySequence(a, this.$$targs$$);
}
Iterable$proto.any = function(/*Callable<Boolean,Element>*/selecting) {
    var iter = this.getIterator();
    var e; while ((e = iter.next()) !== $finished) {
        if (selecting(e)) {
            return true;
        }
    }
    return false;
}
Iterable$proto.$every = function(/*Callable<Boolean,Element>*/selecting) {
    var iter = this.getIterator();
    var e; while ((e = iter.next()) !== $finished) {
        if (!selecting(e)) {
            return false;
        }
    }
    return true;
}
Iterable$proto.skipping = function(skip) {
    function skip$iter(iter,skip){
        var $cmp$=new skip$iter.$$;
        Basic($cmp$);
        $cmp$.iter=iter;
        $cmp$.skip=skip;
        $cmp$.getIterator=function(){
            var iter = this.iter.getIterator();
            for (var i=0; i < this.skip; i++) {
                iter.next();
            }
            return iter;
        };
        return $cmp$;
    }
    initTypeProto(skip$iter, 'ceylon.language::SkipIterable', Basic, Iterable);
    return skip$iter(this,skip);
}
Iterable$proto.taking = function(take) {
    if (take <= 0) return empty;
    var iter = this;
    return Comprehension(function() {
        var it = iter.getIterator();
        var i = 0;
        return function() {
            if (i >= take) {return $finished;}
            ++i;
            return it.next();
        }
    });
}
Iterable$proto.by = function(step) {
    if (step == 1) return this;
    if (step < 1) throw Exception(String$("Step must be positive"));
    var iter = this;
    return Comprehension(function() {
        var it = iter.getIterator();
        return function() {
            var e = it.next();
            for (var i=1; i<step && (it.next()!==$finished); i++);
            return e;
        }
    });
}
Iterable$proto.count = function(sel) {
    var c = 0;
    var iter = this.getIterator();
    var e; while ((e = iter.next()) !== $finished) {
        if (sel(e)) c++;
    }
    return c;
}
Iterable$proto.getCoalesced = function() {
    var iter = this;
    return Comprehension(function() {
        var it = iter.getIterator();
        return function() {
            var e;
            while ((e = it.next()) === null);
            return e;            
        }
    });
}
Iterable$proto.getIndexed = function() {
    var iter = this;
    return Comprehension(function() {
        var it = iter.getIterator();
        var idx = 0;
        return function() {
            var e;
            while ((e = it.next()) === null) {idx++;}
            return e === $finished ? e : Entry(idx++, e);
        }
    });
}
Iterable$proto.getLast = function() {
    var iter = this.getIterator();
    var l=null;
    var e; while ((e = iter.next()) !== $finished) {
        l=e;
    }
    return l;
}
Iterable$proto.collect = function(collecting) {
    return this.$map(collecting).getSequence();
}
Iterable$proto.select = function(selecting) {
    return this.$filter(selecting).getSequence();
}
Iterable$proto.group = function(grouping, $$$mptypes) {
    var map = HashMap();
    var it = this.getIterator();
    var elem;
    var newSeq = ArraySequence([], this.$$targs$$);
    while ((elem=it.next()) !== $finished) {
        var key = grouping(elem);
        var seq = map.put(Entry(key, newSeq, [$$$mptypes[0], {t:Sequence, a:this.$$targs$$}]), true);
        if (seq === null) {
            seq = newSeq;
            newSeq = ArraySequence([], this.$$targs$$);
        }
        seq.push(elem);
    }
    return map;
}
Iterable$proto.chain = function(other, $$$mptypes) {
    return ChainedIterable(this, other, [this.$$targs$$[0], $$$mptypes[0]]);
}
Iterable$proto.getSize = function() {
    return this.count(function() { return true; });
}
exports.Iterable=Iterable;

function ChainedIterable(first, second, $$targs$$, chained) {
    if (chained===undefined) {chained = new ChainedIterable.$$;}
    Basic(chained);
    chained.first = first;
    chained.second = second;
    chained.$$targs$$=$$targs$$;
    return chained;
}
initTypeProto(ChainedIterable, "ceylon.language::ChainedIterable",
        Basic, Iterable);
var ChainedIterable$proto = ChainedIterable.$$.prototype;
ChainedIterable$proto.getIterator = function() {
    return ChainedIterator(this.first, this.second);
}

function toTuple(iterable) {
  var seq = iterable.getSequence();
  return Tuple(seq.getFirst(), seq.getRest().getSequence(),
    [seq.$$targs$$[0], seq.$$targs$$[0], {t:Sequential, a:seq.$$targs$$}]);
}
exports.toTuple=toTuple;
